Modules/_asynciomodule.c

```diff
@@ -1939,84 +1934,63 @@ enter_task(asyncio_state *state, PyObject *loop, PyObject *task)
         Py_DECREF(item);
         return -1;
     }
-    Py_DECREF(item);
-    return 0;
-}
-
-static int
-err_leave_task(PyObject *item, PyObject *task)
-{
-    PyErr_Format(
-        PyExc_RuntimeError,
-        "Leaving task %R does not match the current task %R.",
-        task, item);
-    return -1;
-}
 
-static int
-leave_task_predicate(PyObject *item, void *task)
-{
-    if (item != task) {
-        return err_leave_task(item, (PyObject *)task);
+    if (PyObject_SetAttr(loop, &_Py_ID(_current_task), task) < 0) {
+        return -1;
     }
-    return 1;
+    return 0;
 }
 
 static int
 leave_task(asyncio_state *state, PyObject *loop, PyObject *task)
-/*[clinic end generated code: output=0ebf6db4b858fb41 input=51296a46313d1ad8]*/
 {
-    int res = _PyDict_DelItemIf(state->current_tasks, loop,
-                                leave_task_predicate, task);
-    if (res == 0) {
-        // task was not found
-        return err_leave_task(Py_None, task);
+    PyObject *item;
+    if (PyObject_GetOptionalAttr(loop, &_Py_ID(_current_task), &item) < 0) {
+        return -1;
     }
-    return res;
-}
 
-static PyObject *
-swap_current_task_lock_held(PyDictObject *current_tasks, PyObject *loop,
-                            Py_hash_t hash, PyObject *task)
-{
-    PyObject *prev_task;
-    if (_PyDict_GetItemRef_KnownHash_LockHeld(current_tasks, loop, hash, &prev_task) < 0) {
-        return NULL;
+    if (item == NULL || item == Py_None) {
+        // current task is not set
+        PyErr_Format(PyExc_RuntimeError,
+                     "Leaving task %R does not match the current task %R.",
```

Comment:
Having "None" in the message could be a bit confusing. Perhaps something like "Leaving task %R does not match the current task (None)" or "Unexpected leaving task %R"?